azure-storage-common/src/Common/Internal/Authentication/MSIAuthScheme.php

```diff
@@ -0,0 +1,258 @@
+        $canonicalizedResource .= parse_url($url, PHP_URL_PATH);
+
+        // 3. The query string should include the question mark and the comp
+        //    parameter (for example, ?comp=metadata). No other parameters should
+        //    be included on the query string.
+        if (array_key_exists(Resources::QP_COMP, $queryParams)) {
+            $canonicalizedResource .= '?' . Resources::QP_COMP . '=';
+            $canonicalizedResource .= $queryParams[Resources::QP_COMP];
+        }
+
+        return $canonicalizedResource;
+    }
+
+    /**
+     * Computes canonicalized resources from URL.
+     *
+     * @param string $url         request url.
+     * @param array  $queryParams request query variables.
+     *
+     * @see Constructing the Canonicalized Resource String section at
+     *      http://msdn.microsoft.com/en-us/library/windowsazure/dd179428.aspx
+     *
+     * @return string
+     */
+    protected function computeCanonicalizedResource($url, $queryParams)
+    {
+        $queryParams = array_change_key_case($queryParams);
+
+        // 1. Beginning with an empty string (""), append a forward slash (/),
+        //    followed by the name of the account that owns the accessed resource.
+        $canonicalizedResource = '/' . $this->accountName;
+
+        // 2. Append the resource's encoded URI path, without any query parameters.
+        $canonicalizedResource .= parse_url($url, PHP_URL_PATH);
+
+        // 3. Retrieve all query parameters on the resource URI, including the comp
+        //    parameter if it exists.
+        // 4. Sort the query parameters lexicographically by parameter name, in
+        //    ascending order.
+        if (count($queryParams) > 0) {
+            ksort($queryParams);
+        }
+
+        // 5. Convert all parameter names to lowercase.
+        // 6. URL-decode each query parameter name and value.
+        // 7. Append each query parameter name and value to the string in the
+        //    following format:
+        //      parameter-name:parameter-value
+        // 9. Group query parameters
+        // 10. Append a new line character (\n) after each name-value pair.
+        foreach ($queryParams as $key => $value) {
+            // $value must already be ordered lexicographically
+            // See: ServiceRestProxy::groupQueryValues
+            $canonicalizedResource .= "\n" . $key . ':' . $value;
+        }
+
+        return $canonicalizedResource;
+    }
+
+    /**
+     * Adds authentication header to the request headers.
+     *
+     * @param  \GuzzleHttp\Psr7\Request $request HTTP request object.
+     *
+     * @abstract
+     *
+     * @return \GuzzleHttp\Psr7\Request
+     */
+    public function signRequest(Request $request)
+    {
+        $requestHeaders = HttpFormatter::formatHeaders($request->getHeaders());
+
+        $signedKey = $this->getAuthorizationHeader();
+
+        return $request->withHeader(Resources::AUTHENTICATION, $signedKey);
```

Comment:
Normally, to support OAuth, we need to enable the user to refresh the token themselves or supply a functionality to refresh the token using customized refresh mechanism with the frequency also specified by customer. This will not be compatible with our design.

Reply:
Based on the design of the other SDKs, we don't need to implement the refresh mechanism but only allow user to update the token manually, right ? E.g. user needs to setup a timer and refresh the token whenever the timer expires. For instance in the SDK v10, there is a TokenCredential class with a updateToken method on it to refresh the token via a different thread.

Reply:
Since PHP does not come natively with a multi-threading support, it is difficult to support refresh the token proactively. That is why it is better to expose an API to the customer to set the OAuth token directly.

Reply:
Also, there should be unit test case to make sure that the out going request contains this header when OAuth authentication is selected, together with end-to-end scenario covered by target testing.